File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/payload/TrackerPayloadTest.java
/*
 * Copyright (c) 2015-2022 Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */

package com.snowplowanalytics.snowplow.payload;

import android.test.AndroidTestCase;
import android.util.Base64;

import org.json.JSONException;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.util.HashMap;
import java.util.Map;

public class TrackerPayloadTest extends AndroidTestCase {

    TrackerPayload payload;

    public void setUp() {
        payload = new TrackerPayload();
    }

    public HashMap<String,Object> createTestMap() {
        HashMap<String,Object> map = new HashMap<>();
        map.put("a", "string");
        map.put("b", "");
        map.put("c", null);
        return map;
    }

    public void testAddKeyWhenValue() {
        payload.add("a", "string");
        assertEquals("string", payload.getMap().get("a"));
        payload.add("a", 123);
        assertEquals(123, payload.getMap().get("a"));
    }

    public void testNotAddKeyWhenNullOrEmptyValue() {
        payload.add("a", null);
        assertFalse(payload.getMap().containsKey("a"));
        payload.add("a", "");
        assertFalse(payload.getMap().containsKey("a"));
    }

    public void testRemoveKeyWhenNullOrEmptyValue() {
        payload.add("a", "string");
        payload.add("a", "");
        assertFalse(payload.getMap().containsKey("a"));
        payload.add("a", 123);
        payload.add("a", null);
        assertFalse(payload.getMap().containsKey("a"));
    }
    
    public void testAddMapWithoutNullValueEntries() {
        Map<String,Object> testMap = new HashMap<String,Object>(createTestMap());
        payload.addMap(testMap);
        assertEquals("string", payload.getMap().get("a"));
        assertEquals("", payload.getMap().get("b"));
        assertFalse(payload.getMap().containsKey("c"));
    }

    public void testDontCrashWhenAddNullMap() {
        payload.addMap(null);
        assertTrue(payload.getMap().isEmpty());
        payload.addMap(null, true, "a", "b");
        assertTrue(payload.getMap().isEmpty());
    }

    public void testAddSimpleMapBase64NoEncode() throws JSONException {
        payload.addMap(createTestMap(), false, "enc", "no_enc");
        Map<String, Object> map = payload.getMap();
        assertFalse(map.containsKey("enc"));
        assertTrue(map.containsKey("no_enc"));
        JSONObject json = new JSONObject((String) map.get("no_enc"));
        assertEquals("string", json.getString("a"));
        assertEquals("", json.getString("b"));
        assertEquals(JSONObject.NULL, json.get("c"));
    }

    public void testAddMapBase64Encoded() throws JSONException {
        payload.addMap(createTestMap(), true, "enc", "no_enc");
        Map<String, Object> map = payload.getMap();
        assertFalse(map.containsKey("no_enc"));
        String base64Json = (String) payload.getMap().get("enc");
        String jsonString = "";
        try {
            jsonString = new String(Base64.decode(base64Json, Base64.DEFAULT), "UTF-8");
        } catch (UnsupportedEncodingException e) {
            e.printStackTrace();
            fail("UnsupportedEncodingException");
        }
        JSONObject json = new JSONObject(jsonString);
        assertEquals("string", json.getString("a"));
        assertEquals("", json.getString("b"));
        assertEquals(JSONObject.NULL, json.get("c"));
    }

    public void testSimplePayloadToString() throws JSONException {
        payload.add("a", "string");
        JSONObject map = new JSONObject(payload.toString());
        assertEquals("string", map.getString("a"));
    }

    public void testComplexPayloadToString() throws JSONException {
        payload.add("a", createTestMap());
        JSONObject map = new JSONObject(payload.toString());
        JSONObject innerMap = map.getJSONObject("a");
        assertEquals("string", innerMap.getString("a"));
    }
}

File: snowplow-tracker/src/androidTest/java/com/snowplowanalytics/snowplow/payload/TrackerPayloadTest.kt
/*
 * Copyright (c) 2015-present Snowplow Analytics Ltd. All rights reserved.
 *
 * This program is licensed to you under the Apache License Version 2.0,
 * and you may not use this file except in compliance with the Apache License Version 2.0.
 * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the Apache License Version 2.0 is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
 */
package com.snowplowanalytics.snowplow.payload

import android.util.Base64
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.snowplowanalytics.core.utils.Util.objectMapToString
import org.json.JSONException
import org.json.JSONObject
import org.junit.Assert
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import java.io.UnsupportedEncodingException

@RunWith(AndroidJUnit4::class)
class TrackerPayloadTest {
    var payload: TrackerPayload? = null
    @Before
    fun setUp() {
        payload = TrackerPayload()
    }

    private fun createTestMap(): HashMap<String, Any?> {
        val map = HashMap<String, Any?>()
        map["a"] = "string"
        map["b"] = ""
        map["c"] = null
        return map
    }

    @Test
    fun testAddKeyWhenValue() {
        payload!!.add("a", "string")
        Assert.assertEquals("string", payload!!.map["a"])
        payload!!.add("a", 123)
        Assert.assertEquals(123, payload!!.map["a"])
    }

    @Test
    fun testNotAddKeyWhenNullOrEmptyValue() {
        payload!!.add("a", null)
        Assert.assertFalse(payload!!.map.containsKey("a"))
        payload!!.add("a", "")
        Assert.assertFalse(payload!!.map.containsKey("a"))
    }

    @Test
    fun testRemoveKeyWhenNullOrEmptyValue() {
        payload!!.add("a", "string")
        payload!!.add("a", "")
        Assert.assertFalse(payload!!.map.containsKey("a"))
        payload!!.add("a", 123)
        payload!!.add("a", null)
        Assert.assertFalse(payload!!.map.containsKey("a"))
    }

    @Test
    fun testAddMapWithoutNullValueEntries() {
        val testMap: Map<String, Any?> = HashMap(createTestMap())
        payload!!.addMap(testMap)
        Assert.assertEquals("string", payload!!.map["a"])
        Assert.assertEquals("", payload!!.map["b"])
        Assert.assertFalse(payload!!.map.containsKey("c"))
    }

    @Test
    @Throws(JSONException::class)
    fun testAddSimpleMapBase64NoEncode() {
        payload!!.addMap(createTestMap(), false, "enc", "no_enc")
        val map: Map<String, Any> = payload!!.map
        Assert.assertFalse(map.containsKey("enc"))
        Assert.assertTrue(map.containsKey("no_enc"))
        val json = JSONObject(map["no_enc"] as String)
        Assert.assertEquals("string", json.getString("a"))
        Assert.assertEquals("", json.getString("b"))
        Assert.assertEquals(JSONObject.NULL, json["c"])
    }

    @Test
    @Throws(JSONException::class)
    fun testAddMapBase64Encoded() {
        payload!!.addMap(createTestMap(), true, "enc", "no_enc")
        val map = objectMapToString(
            payload!!.map
        )
        Assert.assertFalse(map.containsKey("no_enc"))
        val base64Json = payload!!.map["enc"] as String?
        var jsonString = ""
        try {
            jsonString = String(Base64.decode(base64Json, Base64.DEFAULT), Charsets.UTF_8)
        } catch (e: UnsupportedEncodingException) {
            e.printStackTrace()
            Assert.fail("UnsupportedEncodingException")
        }
        val json = JSONObject(jsonString)
        Assert.assertEquals("string", json.getString("a"))
        Assert.assertEquals("", json.getString("b"))
        Assert.assertEquals(JSONObject.NULL, json["c"])
    }

    @Test
    @Throws(JSONException::class)
    fun testSimplePayloadToString() {
        payload!!.add("a", "string")
        val map = JSONObject(payload.toString())
        Assert.assertEquals("string", map.getString("a"))
    }

    @Test
    @Throws(JSONException::class)
    fun testComplexPayloadToString() {
        payload!!.add("a", createTestMap())
        val map = JSONObject(payload.toString())
        val innerMap = map.getJSONObject("a")
        Assert.assertEquals("string", innerMap.getString("a"))
    }
}
